addons/point_of_sale/models/pos_order.py

```diff
@@ -606,19 +605,173 @@ def _prepare_invoice_vals(self):
+            'partner_id': self.partner_id.id,
+            'balance': -total_balance,
+        })
+        return aml_vals_list
+
+    def _create_misc_reversal_move(self, payment_moves):
+        """ Create a misc move to reverse this POS order and "remove" it from the POS closing entry.
+        This is done by taking data from the order and using it to somewhat replicate the resulting entry in order to
+        reverse partially the movements done ine the POS closing entry.
+        """
+        aml_values_list = self._prepare_aml_values_list()
+        for aml_values in aml_values_list:
+            aml_values['balance'] = -aml_values['balance']
+
+        # Make a move with all the lines.
+        reversal_entry = self.env['account.move'].with_context(default_journal_id=self.config_id.journal_id.id).create({
+            'journal_id': self.config_id.journal_id.id,
+            'date': fields.Date.today(),
+            'ref': _('Reversal of POS closing entry %s for %s') % (self.session_move_id.name, self.name),
+            'invoice_line_ids': [Command.create(aml_value) for aml_value in aml_values_list],
+        })
+        reversal_entry._post()
+
+        # Reconcile the new receivable line with the lines from the payment move.
+        pos_account_receivable = self.company_id.account_default_pos_receivable_account_id
+        reversal_entry_receivable = reversal_entry.line_ids.filtered(lambda l: l.account_id == pos_account_receivable)
+        payment_receivable = payment_moves.line_ids.filtered(lambda l: l.account_id == pos_account_receivable)
+        (reversal_entry_receivable | payment_receivable).reconcile()
+
+    def action_pos_order_invoice(self, postponed_generation=False, move_vals=None):
+        if any(order.is_invoiced for order in self):
+            error_action = {
+                'name': _('Invoices'),
+                'res_model': 'account.move',
+                'type': 'ir.actions.act_window',
+                'target': 'current',
+                'view_mode': 'tree,form',
+                'views': [[False, 'list'], [False, 'form']],
+                'domain': [('id', 'in', self.account_move.ids)],
+            }
+            if len(self.account_move) == 1:
+                error_action['views'] = [[False, 'form']]
+                error_action['res_id'] = self.account_move.id
+            raise RedirectWarning(_('An invoice was already generated for some of the selected orders.'), error_action, _('Display the invoice'))
         self.write({'to_invoice': True})
-        res = self._generate_pos_order_invoice()
+        res = self._generate_pos_order_invoice(postponed_generation, move_vals)
         if self.company_id.anglo_saxon_accounting and self.session_id.update_stock_at_closing:
             self._create_order_picking()
         return res
 
-    def _generate_pos_order_invoice(self):
+    def _generate_pos_order_invoice(self, postponed_generation=False, move_vals=None):
```

Comment:
postponed_generation attribute is not needed, you can just check that the session i closed or not.

Reply:
also remove move_vals as postponed_generation